datasource/src/test/scala/quasar/physical/mongo/interpreter/CartesianSpec.scala

```diff
@@ -0,0 +1,103 @@
+import quasar.{ScalarStage, IdStatus}
+
+class CartesianSpec extends Specification {
+  import E.helpers._
+  def evalCartesian(cartouches: Map[CPathField, (CPathField, List[ScalarStage.Focused])]): Option[List[Aggregator]] = {
+    val interpreter = new MongoInterpreter(Version.$objectToArray, "root")
+    Cartesian("root", Version.$objectToArray, cartouches, interpreter)
+  }
+  "empty cartesian should erase everything" >> {
+    val actual = evalCartesian(Map.empty)
+    val expected = Some(List(Aggregator.filter(E.Object("root_cartesian_empty" -> E.Bool(false)))))
+    actual === expected
+  }
+  "example" >> {
+    val cartouches = Map(
+      (CPathField("a"), (CPathField("a"), Nil)),
+
+      (CPathField("ba"), (CPathField("b"), List(
+        ScalarStage.Mask(Map(CPath.Identity -> Set(ColumnType.Array))),
+        ScalarStage.Pivot(IdStatus.ExcludeId, ColumnType.Array)))),
+
+      (CPathField("bm"), (CPathField("b"), List(
+        ScalarStage.Mask(Map(CPath.Identity -> Set(ColumnType.Object))),
+        ScalarStage.Pivot(IdStatus.ExcludeId, ColumnType.Object)))))
+
+    val actual = evalCartesian(cartouches)
+
+    val expected = Some(List(
+      Aggregator.project(E.Object(
+        "a" -> (E.key("root") +/ E.key("a")),
+        "ba" -> (E.key("root") +/ E.key("b")),
+        "bm" -> (E.key("root") +/ E.key("b")))),
+      Aggregator.project(E.Object(
+        "a" -> E.key("a"),
+        "ba" -> cond(
+          equal(typeExpr(E.key("ba")), E.String("array")),
+          E.key("ba"),
+          E.key("ba_non_existent_field")),
+        "bm" -> E.key("bm"))),
+      Aggregator.project(E.Object(
+        "a" -> E.key("a"),
+        "ba" -> E.key("ba"),
+        "bm" -> E.key("bm"),
+        "ba_unwind" -> E.key("ba"))),
+      Aggregator.unwind(E.key("ba_unwind"), "ba_unwind_index"),
+      Aggregator.project(E.Object(
+        "a" -> E.key("a"),
+        "ba" -> E.key("ba_unwind"),
+        "bm" -> E.key("bm"))),
+      Aggregator.project(E.Object(
+        "a" -> E.key("a"),
+        "ba" -> E.key("ba"),
+        "bm" -> cond(
+          equal(typeExpr(E.key("bm")), E.String("object")),
+          E.key("bm"),
+          E.key("bm_non_existent_field")))),
+      Aggregator.project(E.Object(
+        "a" -> E.key("a"),
+        "ba" -> E.key("ba"),
+        "bm" -> E.key("bm"),
+        "bm_unwind" -> E.Object("$objectToArray" -> E.key("bm")))),
+      Aggregator.unwind(E.key("bm_unwind"), "bm_unwind_index"),
+      Aggregator.project(E.Object(
+        "a" -> E.key("a"),
+        "ba" -> E.key("ba"),
+        "bm" -> (E.key("bm_unwind") +/ E.key("v")))),
+      Aggregator.project(E.Object(
+        "root" -> E.Object(
+          "a" -> E.key("a"),
+          "ba" -> E.key("ba"),
+          "bm" -> E.key("bm")))),
+      Aggregator.notNull("root")))
+
+    val expectedDocs = expected map (_ map (_.toDocument))
+    val actualDocs = actual map (_ map (_.toDocument))
+    expectedDocs === actualDocs
```

Comment:
Not asking for change, but it's sometimes nicer to use `beTreeEqual` from `quasar.matchers` to make these assertions